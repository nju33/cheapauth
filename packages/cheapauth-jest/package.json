{
  "private": true,
  "name": "cheapauth-jest",
  "version": "0.0.2-alpha.0",
  "dependencies": {
    "@testing-library/react": "^8.0.1",
    "@types/jest": "26.0.15",
    "@vue/test-utils": "^1.0.0-beta.29",
    "jest": "24.9.0",
    "react-testing-library": "^8.0.1"
  },
  "devDependencies": {
    "typescript": "3.9.7",
    "vue": "^2.6.10"
  },
  "peerDependencies": {
    "vue": "^2.6.0"
  }
}
