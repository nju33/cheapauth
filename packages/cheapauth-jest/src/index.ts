import {RenderResult} from '@testing-library/react';
import Vue from 'vue';
import {Wrapper} from '@vue/test-utils';
import TestId from './constants/testid';

export type CheapauthRenderResult = RenderResult | Wrapper<Vue>;

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toBeTitleTextContent(text: string): R;
      toBeSubmitTextContent(text: string): R;
      toBeErrorMessageTextContent(text: string): R;
      toBeApproved(): R;
    }
  }
}

const isReactTest = (
  renderResult: CheapauthRenderResult,
): renderResult is RenderResult => {
  return Object.prototype.hasOwnProperty.call(renderResult, 'asFragment');
};

const pass = {
  pass: true,
  message: '',
};

const toBeOneTextContent = (testid: TestId) => (
  renderResult: CheapauthRenderResult,
  text: string,
): jest.CustomMatcherResult => {
  if (isReactTest(renderResult)) {
    const {queryByTestId} = renderResult;

    const targetElement = queryByTestId(testid);
    if (targetElement === null) {
      return {
        pass: false,
        message: () => {
          return `expected to have the element \`[data-testid='${testid}']\``;
        },
      };
    }

    if (targetElement.textContent !== text) {
      return {
        pass: false,
        message: () => {
          return `expected to have the element \`[data-testid='${testid}']{${
            targetElement.textContent
          }}\``;
        },
      };
    }
  } else {
    const titleWrapper = renderResult.find(`[data-testid='${testid}']`);
    const expectedText = titleWrapper.text();

    if (!(titleWrapper.exists() && expectedText === text)) {
      return {
        pass: false,
        message: () => {
          return `expected to have the element \`[data-testid='${testid}']{${expectedText}}\``;
        },
      };
    }
  }

  return pass;
};

expect.extend({
  toBeTitleTextContent: toBeOneTextContent(TestId.title),
  toBeSubmitTextContent: toBeOneTextContent(TestId.submit),
  toBeErrorMessageTextContent: toBeOneTextContent(TestId.errorMessage),
  toBeApproved(renderResult: CheapauthRenderResult): jest.CustomMatcherResult {
    if (isReactTest(renderResult)) {
      const targetElement = renderResult.queryByTestId(TestId.root);
      if (targetElement === null) {
        return {
          pass: false,
          message: () => {
            return `expected to have to element \`[data-testid='${
              TestId.root
            }']\``;
          },
        };
      }

      if (targetElement.dataset.approval !== 'true') {
        return {
          pass: false,
          message: () => {
            return `expected to have to element \`[data-testid='${
              TestId.root
            }'][data-approval='true']\``;
          },
        };
      }
    } else {
      const titleWrapper = renderResult.find(`[data-testid='${TestId.root}']`);
      const approval = titleWrapper.attributes()['data-approval'];

      if (!(titleWrapper.exists() && approval === 'true')) {
        return {
          pass: false,
          message: () => {
            return `expected to have to element \`[data-testid='${
              TestId.root
            }'][data-approval='true']\``;
          },
        };
      }
    }

    return pass;
  },
});
