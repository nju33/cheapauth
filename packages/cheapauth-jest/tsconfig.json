{
  "compilerOptions": {
    "target": "es5",
    "module": "esnext",
    "typeRoots": ["./node_modules/@types", "../../node_modules/@types"],
    "jsx": "react",
    "sourceMap": true,
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "downlevelIteration": true
  },
  "include": ["src"],
  "exclude": ["node_modules"]
}
