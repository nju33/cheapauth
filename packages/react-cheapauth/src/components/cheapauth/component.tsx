import React from 'react';
import originalStyled, {
  StyledComponentBase,
  ThemedBaseStyledInterface,
} from 'styled-components';
import {
  CheapauthThemeStruct,
  CheapauthComponentProps,
  CheapauthComponentState,
} from './interfaces';
import Password from '../../entities/password';
import {init, handleAuthentication} from './use-cases';

const component = ({} as unknown) as {
  root: StyledComponentBase<'div', {}>;
};

export class Cheapauth extends React.PureComponent<
  CheapauthComponentProps,
  CheapauthComponentState
> {
  public state: CheapauthComponentState = {
    password: undefined,
    approval: false,
    error: false,
  };

  public inputRef = React.createRef<HTMLInputElement>();

  public componentDidMount() {
    const answer = new Password(this.props.password);

    this.setState({
      password: answer,
      approval: init(answer),
    });

    if (this.inputRef.current !== null) {
      this.inputRef.current.focus();
    }
  }

  public onSubmit = (ev: React.FormEvent): void => {
    ev.preventDefault();

    const input = this.inputRef.current;
    if (input === null) {
      return;
    }

    if (this.state.password === undefined) {
      return;
    }

    const approval = handleAuthentication(this.state.password, input.value);

    if (!approval) {
      this.setState({
        error: true,
        errorMessage: 'パスワードが違います',
        approval: false,
      });
      return;
    }

    this.setState({approval});
  };

  public render(): JSX.Element {
    return (
      <>
        <component.root
          theme={this.props.theme}
          data-approval={this.state.approval}
          data-testid="cheapauth--root"
        >
          <div>
            <section className="section" aria-labelledby="cheapauth">
              <header className="header">
                <h3 className="title" data-testid="cheapauth--title">
                  {this.props.title}
                </h3>
              </header>
              <div>
                <form
                  className="form"
                  data-testid="cheapauth--form"
                  onSubmit={this.onSubmit}
                >
                  <div className="row">
                    <input
                      ref={this.inputRef}
                      type="password"
                      className="input"
                      data-testid="cheapauth--input"
                    />
                    {this.state.error && (
                      <div
                        className="error-message"
                        data-testid="cheapauth--error-message"
                        aria-live="assertive"
                      >
                        <small>{this.state.errorMessage}</small>
                      </div>
                    )}
                  </div>
                  <div className="row">
                    <button className="submit" data-testid="cheapauth--submit">
                      {this.props.submitLabel}
                    </button>
                  </div>
                </form>
              </div>
            </section>
          </div>
        </component.root>
        {this.props.children}
      </>
    );
  }
}

export default Cheapauth;

const styled = originalStyled as ThemedBaseStyledInterface<
  CheapauthThemeStruct
>;
component.root = styled.div`
  position: relative;
  z-index: 999999999;

  &[data-approval='true'] {
    display: none;
  }

  > div {
    position: fixed;
    width: 150vw;
    height: 150vh;
    right: 50%;
    bottom: 50%;
    transform: translate(50%, 50%);
    background: ${props => props.theme.backgroundColor || '#f8f8f8'};
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  .header {
    margin-left: 0.5em;
  }

  .section {
    width: 90vw;
  }

  .title {
    text-align: center;
    margin-bottom: 1.5em;
    font-size: 2em;
  }

  .form {
    display: flex;
    flex-direction: column;
    /* align-items: center; */
  }

  .row {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
  }

  .input {
    font-size: 1em;
    height: 2.5em;
    width: 100%;
    padding: 0.5em 1em;
    box-sizing: border-box;
    border: none;
    box-shadow: 0 0 2px -1px rgba(0, 0, 0, 0.5);
  }

  .error-message {
    position: absolute;
    left: 0;
    bottom: -1.5em;
    color: #cb1b45;
    margin-left: 0.5em;
  }

  .submit {
    font-size: 1em;
    max-width: 55vw;
    width: 100%;
    height: 2.5em;
    padding: 0.5em 1em;
    box-sizing: border-box;
    cursor: pointer;
    border: none;
    background: ${props => props.theme.accentColor || '#2ea9df'};
    margin-top: 2em;
  }

  @media screen and (min-width: 758px) {
    .section {
      width: auto;
    }

    .title {
      text-align: left;
      margin-bottom: 0.5em;
      font-size: 1.2em;
    }

    .form {
      flex-direction: row;
      align-items: flex-start;
    }

    .input {
      width: 30em;
    }

    .submit {
      width: auto;
      margin-top: 0;
      margin-left: 0.5em;
    }
  }
`;
