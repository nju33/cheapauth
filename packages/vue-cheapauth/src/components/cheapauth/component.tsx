import Vue, {CreateElement, VNode} from 'vue';
import Component from 'vue-class-component';
import styled from 'vue-styled-components';
import Password from '../../entities/password';
import {init, handleAuthentication} from './use-cases';

/* eslint-disable @typescript-eslint/no-namespace, @typescript-eslint/no-empty-interface */
declare global {
  namespace JSX {
    // @ts-ignore
    interface Element extends VNode {}
    interface ElementClass extends Vue {}
    interface IntrinsicElements {
      [elem: string]: any; // eslint-disable-line @typescript-eslint/no-explicit-any
    }
  }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const Root = styled.div`
  position: relative;
  z-index: 999999999;

  &[data-approval='true'] {
    display: none;
  }

  > div {
    position: fixed;
    width: 150vw;
    height: 150vh;
    right: 50%;
    bottom: 50%;
    transform: translate(50%, 50%);
    background: #f8f8f8;
    /* background: ${props => props.theme.backgroundColor || '#f8f8f8'}; */
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  [data-class='header'] {
    margin-left: 0.5em;
  }

  [data-class='section'] {
    width: 90vw;
  }

  [data-class='title'] {
    text-align: center;
    margin-bottom: 1.5em;
    font-size: 1.5em;
  }

  [data-class='form'] {
    display: flex;
    flex-direction: column;
    /* align-items: center; */
  }

  [data-class='row'] {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
  }

  [data-class='input'] {
    font-size: 1em;
    height: 2.5em;
    width: 100%;
    padding: 0.5em 1em;
    box-sizing: border-box;
    border: none;
    box-shadow: 0 0 2px -1px rgba(0, 0, 0, 0.5);
  }

  [data-class='error-message'] {
    position: absolute;
    bottom: -1.5em;
    color: #cb1b45;
    margin-left: 0.5em;
  }

  [data-class='submit'] {
    font-size: 1em;
    max-width: 55vw;
    width: 100%;
    height: 2.5em;
    padding: 0.5em 1em;
    box-sizing: border-box;
    cursor: pointer;
    border: none;
    background: #2ea9df;
    /* background: ${props => props.theme.accentColor || '#2ea9df'}; */
    margin-top: 1.5em;
  }

  @media screen and (min-width: 758px) {
    [data-class='section'] {
      width: auto;
    }

    [data-class='title'] {
      text-align: left;
      margin-bottom: 0.5em;
      font-size: 1.2em;
    }

    [data-class='form'] {
      flex-direction: row;
      align-items: flex-start;
    }

    [data-class='input'] {
      width: 30em;
    }

    [data-class='submit'] {
      width: auto;
      margin-top: 0;
      margin-left: 0.5em;
    }
  }
`;

/* eslint-disable new-cap, @typescript-eslint/explicit-member-accessibility */
@Component({
  props: {
    password: {
      type: String,
      required: true,
    },
    title: {
      type: String,
      required: true,
    },
    submitLabel: {
      type: String,
      required: true,
    },
  },
})
export class Cheapauth extends Vue {
  // public answer: Password = new Password(Math.random().toString());
  public approval = false;

  public input = '';

  public error = false;

  public errorMessage = '';

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public onChange = (ev: any) => {
    ev.preventDefault();

    this.input = (ev.target as HTMLInputElement).value;
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public onSubmit(ev: any) {
    ev.preventDefault();

    const approval = handleAuthentication(
      new Password(this.$props.password),
      this.input,
    );
    if (!approval) {
      this.error = true;
      this.errorMessage = 'パスワードが異なります。';
      return;
    }

    this.approval = true;
  }

  public mounted() {
    const answer = new Password(this.$props.password);
    this.approval = init(answer);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public render(h: CreateElement): VNode {
    return (
      <Root data-testid="cheapauth-root" data-approval={this.approval}>
        <div>
          <section data-class="section" aria-labelledby="cheapauth">
            <header data-class="header">
              <h3 data-class="title">{this.$props.title}</h3>
            </header>
            <div>
              <form
                data-class="form"
                data-testid="form"
                onSubmit={this.onSubmit}
              >
                <div data-class="row">
                  <input
                    type="password"
                    data-class="input"
                    data-testid="input"
                    value={this.input}
                    onChange={this.onChange}
                  />
                  {this.error && (
                    <div data-class="error-message" data-testid="error-message">
                      <small>{this.errorMessage}</small>
                    </div>
                  )}
                </div>
                <div data-class="row">
                  <button data-class="submit">{this.$props.submitLabel}</button>
                </div>
              </form>
            </div>
          </section>
        </div>
      </Root>
    );
  }
}
