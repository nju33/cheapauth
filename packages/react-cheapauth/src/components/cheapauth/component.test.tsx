import React from 'react';
import {render, fireEvent, cleanup, RenderResult} from '@testing-library/react';
import 'jest-dom/extend-expect';
import {Cheapauth} from './component';
import 'cheapauth-jest/extend-expect';

describe('Cheapauth', () => {
  afterEach(cleanup);

  const password = 'password';
  const title = 'タイトル';
  const submitLabel = 'ログイン';

  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <Cheapauth password={password} title={title} submitLabel={submitLabel} />,
    );
  });

  it('is rendered with text of props into right place', () => {
    expect(renderResult).toBeTitleTextContent(title);
    expect(renderResult).toBeSubmitTextContent(submitLabel);
  });

  it('is input password is incorrect', () => {
    const incorrectPassword = 'passworddd';

    expect(renderResult.container.firstChild).toMatchSnapshot();
    expect(renderResult).not.toBeApproved();

    // to fire the submit event
    const input = renderResult.queryByTestId('cheapauth--input');
    (input as HTMLInputElement).value = incorrectPassword;
    fireEvent.submit(renderResult.queryByTestId(
      'cheapauth--form',
    ) as HTMLFormElement);

    expect(renderResult.container.firstChild).toMatchSnapshot();
    expect(renderResult).toBeErrorMessageTextContent('パスワードが違います');
    expect(renderResult).not.toBeApproved();
  });

  test('is input password is correct', () => {
    expect(renderResult.container.firstChild).toMatchSnapshot();
    expect(renderResult).not.toBeApproved();

    // to fire the submit event
    const input = renderResult.queryByTestId('cheapauth--input');
    (input as HTMLInputElement).value = password;
    fireEvent.submit(renderResult.queryByTestId(
      'cheapauth--form',
    ) as HTMLFormElement);

    expect(renderResult.container.firstChild).toMatchSnapshot();
    expect(renderResult).toBeApproved();
  });
});
