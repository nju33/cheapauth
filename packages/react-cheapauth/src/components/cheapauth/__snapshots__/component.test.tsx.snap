// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cheapauth is input password is correct 1`] = `
<div
  class="sc-bdVaJa fAKkVW"
  data-approval="false"
  data-testid="cheapauth--root"
>
  <div
    class="cheapauth--document"
  >
    <section
      aria-labelledby="cheapauth"
      class="cheapauth--section"
    >
      <header
        class="cheapauth--header"
      >
        <h3
          class="cheapauth--title"
          data-testid="cheapauth--title"
        >
          タイトル
        </h3>
      </header>
      <div>
        <form
          class="cheapauth--form"
          data-testid="cheapauth--form"
        >
          <div
            class="cheapauth--row"
          >
            <input
              class="cheapauth--input"
              data-testid="cheapauth--input"
              type="password"
            />
          </div>
          <div
            class="cheapauth--row"
          >
            <button
              class="cheapauth--submit"
              data-testid="cheapauth--submit"
            >
              ログイン
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>
</div>
`;

exports[`Cheapauth is input password is correct 2`] = `
<div
  class="sc-bdVaJa fAKkVW"
  data-approval="true"
  data-testid="cheapauth--root"
>
  <div
    class="cheapauth--document"
  >
    <section
      aria-labelledby="cheapauth"
      class="cheapauth--section"
    >
      <header
        class="cheapauth--header"
      >
        <h3
          class="cheapauth--title"
          data-testid="cheapauth--title"
        >
          タイトル
        </h3>
      </header>
      <div>
        <form
          class="cheapauth--form"
          data-testid="cheapauth--form"
        >
          <div
            class="cheapauth--row"
          >
            <input
              class="cheapauth--input"
              data-testid="cheapauth--input"
              type="password"
            />
          </div>
          <div
            class="cheapauth--row"
          >
            <button
              class="cheapauth--submit"
              data-testid="cheapauth--submit"
            >
              ログイン
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>
</div>
`;

exports[`Cheapauth is input password is incorrect 1`] = `
<div
  class="sc-bdVaJa fAKkVW"
  data-approval="false"
  data-testid="cheapauth--root"
>
  <div
    class="cheapauth--document"
  >
    <section
      aria-labelledby="cheapauth"
      class="cheapauth--section"
    >
      <header
        class="cheapauth--header"
      >
        <h3
          class="cheapauth--title"
          data-testid="cheapauth--title"
        >
          タイトル
        </h3>
      </header>
      <div>
        <form
          class="cheapauth--form"
          data-testid="cheapauth--form"
        >
          <div
            class="cheapauth--row"
          >
            <input
              class="cheapauth--input"
              data-testid="cheapauth--input"
              type="password"
            />
          </div>
          <div
            class="cheapauth--row"
          >
            <button
              class="cheapauth--submit"
              data-testid="cheapauth--submit"
            >
              ログイン
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>
</div>
`;

exports[`Cheapauth is input password is incorrect 2`] = `
<div
  class="sc-bdVaJa fAKkVW"
  data-approval="false"
  data-testid="cheapauth--root"
>
  <div
    class="cheapauth--document"
  >
    <section
      aria-labelledby="cheapauth"
      class="cheapauth--section"
    >
      <header
        class="cheapauth--header"
      >
        <h3
          class="cheapauth--title"
          data-testid="cheapauth--title"
        >
          タイトル
        </h3>
      </header>
      <div>
        <form
          class="cheapauth--form"
          data-testid="cheapauth--form"
        >
          <div
            class="cheapauth--row"
          >
            <input
              class="cheapauth--input"
              data-testid="cheapauth--input"
              type="password"
            />
            <div
              aria-live="assertive"
              class="cheapauth--error-message"
              data-testid="cheapauth--error-message"
            >
              <small>
                パスワードが違います
              </small>
            </div>
          </div>
          <div
            class="cheapauth--row"
          >
            <button
              class="cheapauth--submit"
              data-testid="cheapauth--submit"
            >
              ログイン
            </button>
          </div>
        </form>
      </div>
    </section>
  </div>
</div>
`;
